File: easy/linked_list_cycle/linked_list_cycle.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

    def push(self, x):
        new_node = ListNode(x)
        self.next = new_node
        return new_node

    def print(self):
        node = self
        while node != None:
            print(node.val)
            node = node.next


def array_to_linked_list(arr, cycle_start_pos):
    if len(arr) < 1:
        return None

    head = ListNode(arr[0])
    start, tail = head, head
    for val in arr[1:]:
        new_node = ListNode(val)
        tail.next = new_node
        tail = new_node

    if cycle_start_pos >= 0:
        for i in range(0, cycle_start_pos):
            start = start.next
        tail.next = start

    return head


class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        slow, fast = head, head
        while slow and fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True

        return False


s = Solution()
head = array_to_linked_list([3, 2, 0, -4], 1)
print("Solution 1 : ", s.hasCycle(head))

head = array_to_linked_list([1, 2], 0)
print("Solution 2 : ", s.hasCycle(head))

head = array_to_linked_list([1], -1)
print("Solution 3 : ", s.hasCycle(head))

head = array_to_linked_list([], -1)
print("Solution 4 : ", s.hasCycle(head))

File: easy/linked_list/linked_list_cycle/linked_list_cycle.go
package main

import "fmt"

type Node struct {
	Val  int
	Next *Node
}

func (n *Node) push(val int) *Node {
	newNode := &Node{
		Val:  val,
		Next: nil,
	}
	n.Next = newNode
	n = newNode
	return n
}

func arrayToLinkedList(arr []int, cyclePos int) *Node {
	head := &Node{
		Val:  arr[0],
		Next: nil,
	}
	cycleStart, tail := head, head

	for _, v := range arr[1:] {
		tail = tail.push(v)
	}

	if cyclePos >= 0 {
		for i := 0; i < cyclePos; i++ {
			cycleStart = cycleStart.Next
		}
		tail.Next = cycleStart
	}

	// cur := head
	// for i := 0; i < len(arr); i++ {
	// 	fmt.Println("i = ", i, "val = ", cur.Val)
	// 	cur = cur.Next
	// }
	// fmt.Println("total = ", len(arr))

	return head
}

func hasCycle(head *Node) bool {
	if head == nil || head.Next == nil {
		return false
	}

	slow, fast := head, head.Next
	for slow != fast {
		if fast == nil || fast.Next == nil {
			return false
		}
		slow = slow.Next
		fast = fast.Next.Next
	}

	return true
}

func main() {
	list := arrayToLinkedList([]int{3, 2, 0, -4}, 1)
	fmt.Println("Solution 1 : ", hasCycle(list))

	list = arrayToLinkedList([]int{1, 2}, 0)
	fmt.Println("Solution 2 : ", hasCycle(list))

	list = arrayToLinkedList([]int{1}, -1)
	fmt.Println("Solution 3 : ", hasCycle(list))
}
